File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/inheritance/closure_signature_override/closure_signature_override.swift
import Inheritance
import Testing

@Test
func swiftCanOverrideKotlinMethodTakingClosure() throws {
    // The closure Kotlin passes in `callApply` reaches the Swift override as a Swift closure, and calling
    // it from Swift dispatches back into Kotlin.
    class Doubling: Transformer {
        override func apply(input: Int32, transform: @escaping (Int32) -> Int32) -> Int32 {
            return transform(input) * 2
        }
    }

    #expect(callApply(t: Doubling(), input: 4) == 80)
    // The Kotlin implementation is untouched.
    #expect(callApply(t: Transformer(), input: 4) == 41)
    // Direct Swift dispatch, with a Swift-made closure.
    #expect(Doubling().apply(input: 3, transform: { $0 + 1 }) == 8)
}

@Test
func swiftCanOverrideKotlinMethodReturningClosure() throws {
    // A closure produced by Swift travels into Kotlin, which stores and invokes it.
    class Multiplying: Transformer {
        override func makeAdder(addend: Int32) -> (Int32) -> Int32 {
            return { $0 * addend }
        }
    }

    #expect(callMakeAdder(t: Multiplying(), addend: 3, input: 5) == 15)
    #expect(callMakeAdder(t: Transformer(), addend: 3, input: 5) == 8)
}

@Test
func swiftOverrideOfMethodTakingClosureTakingClosure() throws {
    class Nested: Transformer {
        override func applyTwice(input: Int32, transform: @escaping (@escaping (Int32) -> Int32) -> Int32) -> Int32 {
            return transform({ $0 + 1 }) - input
        }
    }

    // Kotlin passes `{ f -> f(3) }`; the Swift override calls it with its own `{ $0 + 1 }`.
    #expect(callApplyTwice(t: Nested(), input: 10) == -6)
    #expect(callApplyTwice(t: Transformer(), input: 10) == 16)
}

@Test
func swiftOverrideOfMethodWithClosureOverReferenceTypes() throws {
    class Bracketing: Transformer {
        override func describe(name: String, format: @escaping (String) -> String) -> String {
            return format(name) + "!"
        }
    }

    #expect(callDescribe(t: Bracketing(), name: "kotlin") == "[kotlin]!")
    #expect(callDescribe(t: Transformer(), name: "kotlin") == "[kotlin]")
}

@Test
func swiftOverrideOfMethodWithVoidClosure() throws {
    class Repeating: Transformer {
        override func runAction(action: @escaping () -> Void) -> Void {
            action()
            action()
        }
    }

    actionRuns = 0
    callRunAction(t: Repeating())
    #expect(actionRuns == 2)

    actionRuns = 0
    callRunAction(t: Transformer())
    #expect(actionRuns == 1)
}

@Test
func swiftOverrideOfMethodWithOptionalClosure() throws {
    class Optionally: Transformer {
        override func applyOptional(input: Int32, transform: ((Int32) -> Int32)?) -> Int32 {
            guard let transform else { return -2 }
            return transform(input) + 1
        }
    }

    #expect(callApplyOptional(t: Optionally(), input: 1, pass: true) == 102)
    #expect(callApplyOptional(t: Optionally(), input: 1, pass: false) == -2)
    #expect(callApplyOptional(t: Transformer(), input: 1, pass: true) == 101)
    #expect(callApplyOptional(t: Transformer(), input: 1, pass: false) == -1)
}

@Test
func swiftOnlyImplementationOfInterfaceWithClosures() throws {
    // A pure Swift conformer: Kotlin-side dispatch can only reach it through reverse bridges.
    class SwiftProducer: ProducerBase, Producer {
        func produce() -> () -> String {
            return { "from swift" }
        }

        func consume(block: @escaping (String) -> Void) {
            block("swift value")
        }
    }

    let producer = SwiftProducer()
    #expect(callProduce(p: producer) == "from swift")
    callConsume(p: producer, value: "ignored")
    #expect(lastConsumed == "swift value")
}

File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/inheritance/closure_signature_override/closure_signature_override.kt
// KIND: STANDALONE
// FREE_COMPILER_ARGS: -opt-in=kotlin.native.internal.InternalForKotlinNative
// MODULE: Inheritance
// FILE: closure_signature_override.kt

// Cross-language overriding of members that take or return a closure. A reverse bridge flips every
// conversion of the forward bridge, so a closure parameter travels Kotlin -> Swift and a closure
// return value travels Swift -> Kotlin - the direction opposite to the forward bridge.

open class Transformer {
    open fun apply(input: Int, transform: (Int) -> Int): Int = transform(input) + 1

    // Returned closure: the Swift override produces it, Kotlin invokes it.
    open fun makeAdder(addend: Int): (Int) -> Int = { it + addend }

    // Closure that itself takes a closure: nesting flips direction again.
    open fun applyTwice(input: Int, transform: ((Int) -> Int) -> Int): Int = transform { it * 2 } + input

    open fun describe(name: String, format: (String) -> String): String = format(name)

    open fun runAction(action: () -> Unit): Unit = action()

    open fun applyOptional(input: Int, transform: ((Int) -> Int)?): Int = transform?.invoke(input) ?: -1
}

fun callApply(t: Transformer, input: Int): Int = t.apply(input) { it * 10 }
fun callMakeAdder(t: Transformer, addend: Int, input: Int): Int = t.makeAdder(addend)(input)
fun callApplyTwice(t: Transformer, input: Int): Int = t.applyTwice(input) { f -> f(3) }
fun callDescribe(t: Transformer, name: String): String = t.describe(name) { "[$it]" }

var actionRuns: Int = 0
fun callRunAction(t: Transformer) {
    t.runAction { actionRuns += 1 }
}

fun callApplyOptional(t: Transformer, input: Int, pass: Boolean): Int =
    t.applyOptional(input, if (pass) ({ v: Int -> v + 100 }) else null)

interface Producer {
    fun produce(): () -> String
    fun consume(block: (String) -> Unit)
}

// A base without `Producer` in its Kotlin hierarchy: the Swift conformer below can only be reached
// through reverse bridges. A bare `KotlinBase` conformer cannot be constructed yet, see KT-88251.
open class ProducerBase

fun callProduce(p: Producer): String = p.produce()()
fun callConsume(p: Producer, value: String) = p.consume { received -> lastConsumed = received }
var lastConsumed: String = ""
